std/conv.d

```diff
@@ -2816,342 +2828,82 @@ if (isInputRange!Source && isSomeChar!(ElementType!Source) && !is(Source == enum
-                    // base 16: Y = ... + y3*16^3 + y2*16^2 + y1*16^1 + y0*16^0
-                    msdec = msdec * 16 + i;
-                    // ignore leading zeros
-                    if (msdec)
-                        ndigits++;
-                }
-                // All 64 bits of the long have been filled in now
-                else if (ndigits == 16)
-                {
-                    while (msdec >= 0)
-                    {
-                        exp--;
-                        msdec <<= 1;
-                        i <<= 1;
-                        if (i & 0x10)
-                            msdec |= 1;
-                    }
-                    guard = i << 4;
-                    ndigits++;
-                    exp += 4;
-                }
-                else
-                {
-                    guard |= i;
-                    exp += 4;
-                }
-                exp -= dot;
-                p.popFront();
-                if (p.empty)
-                    break;
-                i = p.front;
-                if (i == '_')
-                {
-                    p.popFront();
-                    if (p.empty)
-                        break;
-                    i = p.front;
-                }
-            }
-            if (i == '.' && !dot)
-            {
-                p.popFront();
-                dot = 4;
-            }
-            else
-                break;
-        }
-
-        // Round up if (guard && (sticky || odd))
-        if (guard & 0x80 && (guard & 0x7F || msdec & 1))
-        {
-            msdec++;
-            if (msdec == 0)                 // overflow
-            {
-                msdec = 0x8000000000000000L;
-                exp++;
-            }
-        }
-
-        // Have we seen any mantissa digits so far?
-        enforce(anydigits, bailOut());
-        enforce(!p.empty && (p.front == 'p' || p.front == 'P'),
-                bailOut("Floating point parsing: exponent is required"));
-        char sexp;
-        int e;
-
-        sexp = 0;
-        p.popFront();
-        if (!p.empty)
-        {
-            switch (p.front)
-            {
-                case '-':    sexp++;
-                             goto case;
-                case '+':    p.popFront(); enforce(!p.empty,
-                                new ConvException("Error converting input"~
-                                " to floating point"));
-                             break;
-                default: {}
-            }
-        }
-        ndigits = 0;
-        e = 0;
-        while (!p.empty && isDigit(p.front))
-        {
-            if (e < 0x7FFFFFFF / 10 - 10) // prevent integer overflow
-            {
-                e = e * 10 + p.front - '0';
-            }
-            p.popFront();
-            ndigits = 1;
-        }
-        exp += (sexp) ? -e : e;
-        enforce(ndigits, new ConvException("Error converting input"~
-                        " to floating point"));
+    enum { hex, decimal }
```

Comment:
Ran into an issue when I originally tried to use a named `enum` for the floating-point format, but had to change it like this because it wouldn't compile:
```
diff --git a/std/conv.d b/std/conv.d
index 2b37dfde2..931d0241c 100644
--- a/std/conv.d
+++ b/std/conv.d
@@ -2850,12 +2850,12 @@ if (isInputRange!Source && isSomeChar!(ElementType!Source) && !is(Source == enum
     ulong msscale = 1;
     bool sawDigits;
 
-    enum FloatFormat { Hex, Decimal }
+    enum { hex, decimal }
 
     // sets msdec, lsdec/msscale, and sawDigits by parsing the mantissa digits
-    void parseDigits(FloatFormat digitType)()
+    void parseDigits(alias FloatFormat)()
     {
-        static if (digitType == FloatFormat.Hex)
+        static if (FloatFormat == hex)
         {
             enum uint base = 16;
             enum ulong msscaleMax = 0x1000_0000_0000_0000UL; // largest power of 16 a ulong holds
@@ -2869,7 +2869,7 @@ if (isInputRange!Source && isSomeChar!(ElementType!Source) && !is(Source == enum
             alias convertDigit = (int x) => isAlpha(x) ? ((x & ~0x20) - ('A' - 10)) : x - '0';
             sawDigits = false;
         }
-        else static if (digitType == FloatFormat.Decimal)
+        else static if (FloatFormat == decimal)
         {
             enum uint base = 10;
             enum ulong msscaleMax = 10_000_000_000_000_000_000UL; // largest power of 10 a ulong holds
@@ -2929,9 +2929,9 @@ if (isInputRange!Source && isSomeChar!(ElementType!Source) && !is(Source == enum
     }
 
     if (isHex)
-        parseDigits!FloatFormat.Hex;
+        parseDigits!hex;
     else
-        parseDigits!FloatFormat.Decimal;
+        parseDigits!decimal;
 
     // Have we seen any mantissa digits so far?
     enforce(sawDigits, bailOut("no digits seen"));
```
I looked it up in bugzilla, probably some variation on [bug 13768](https://issues.dlang.org/show_bug.cgi?id=13768), and found [this workaround](https://github.com/libmir/mir-algorithm/blob/master/source/mir/ndslice/slice.d#L198) in [`mir-algorithm`](https://github.com/libmir/mir-algorithm/blob/master/source/mir/ndslice/slice.d#L388), ie if I added `alias fooDec = FloatFormat.Decimal;` and parametrized on that `parseDigits!fooDec;`, it works. Too much boilerplate at that point, so just used the anonymous enums.